### PR TITLE
blobs: omit redundant sha256 digest

### PR DESCRIPTION
**Pull request by danielnorberg**
Currently when uploading large blobs for volume put we hash the file twice:

1. First in `_get_file_upload_spec`
2. Then again in `blob_upload_file`

This PR passes in the initial digest from the file spec in to `blob_upload_file` so that it can skip the second digest. Note that `blob_upload_file` still computes an md5 digest for use in the presigned upload url. We could possibly also remove that, but that requires server side changes as well. Future work.

### CODE REVIEW

File: modal/_utils/blob_utils.py [hunk 1]
```diff
@@ -385,6 +385,7 @@ def _get_file_upload_spec(
         if size >= LARGE_FILE_LIMIT:
             use_blob = True
             content = None
+            sha256_hex = hashlib.file_digest(fp, "sha256")
```
Comment (danielnorberg) [inline modal/_utils/blob_utils.py:388]: derp

File: modal/_utils/hash_utils.py [hunk 2]
```diff
@@ -49,11 +49,18 @@ class UploadHashes:
     sha256_base64: str
 
 
-def get_upload_hashes(data: Union[bytes, BinaryIO]) -> UploadHashes:
+def get_upload_hashes(data: Union[bytes, BinaryIO], sha256_hex: Optional[str] = None) -> UploadHashes:
     md5 = hashlib.md5()
```
Comment (danielnorberg) [inline modal/_utils/hash_utils.py:53]: We could also compute the md5 digest in parallel with the first sha256 digest, on a separate thread. Future work.

**Review by freider — APPROVED** [review]
Nice find!

**Review by gimaker — APPROVED** [review]
:rocket:

**Comment by danielnorberg** [comment]
### Before
```
[ThreadPoolExecutor-0_0] 2024-12-13T10:44:22+0100 get_sha256_hex took 1.972s
[Thread-1 (thread_inner)] 2024-12-13T10:44:30+0100 get_upload_hashes took 8.001s
```

### After
```
[ThreadPoolExecutor-0_0] 2024-12-13T10:44:57+0100 get_sha256_hex took 2.057s
[Thread-1 (thread_inner)] 2024-12-13T10:45:03+0100 get_upload_hashes took 6.289s (['md5'])
```

It seems that unfortunately the remaining md5 hash is very expensive. We should probably omit that and only use the sha256 hash.

**Pull request merged**